crates/storage/nippy-jar/src/reader.rs

```diff
@@ -0,0 +1,244 @@
+use crate::{NippyJarError, OFFSETS_FILE_EXTENSION};
+use memmap2::Mmap;
+use std::{
+    fs::File,
+    io::{Read, Seek, SeekFrom},
+    ops::Range,
+    path::Path,
+};
+
+/// Uses [`FileDataReader`] due to Windows-specific issues with mmap, such as file access locks and
+/// poor handling of file changes.
+#[cfg(windows)]
+pub type DefaultDataReader = FileDataReader;
+
+/// Uses [`MmapDataReader`] for efficient zero-copy read access.
+#[cfg(not(windows))]
+pub type DefaultDataReader = MmapDataReader;
+
+/// Trait defining shared methods for reading data.
+pub trait DataReader {
+    /// Reads the respective data and offsets file and returns [`DataReader`].
+    fn new(path: impl AsRef<Path>) -> Result<Self, NippyJarError>
+    where
+        Self: Sized;
+
+    /// Returns the offset for the requested data index.
+    fn offset(&self, index: usize) -> Result<u64, NippyJarError>;
+
+    /// Returns the offset for the requested data index starting from the end.
+    fn reverse_offset(&self, index: usize) -> Result<u64, NippyJarError>;
+
+    /// Returns the total number of offsets in the file.
+    ///
+    /// The size of one offset is determined by the file itself.
+    fn offsets_count(&self) -> Result<usize, NippyJarError>;
+
+    /// Reads one offset-sized (determined by the offset file) `u64` at the provided index.
+    fn offset_at(&self, index: usize) -> Result<u64, NippyJarError>;
+
+    /// Returns number of bytes that represent one offset.
+    fn offset_size(&self) -> u8;
+
+    /// Whether this data reader supports [`Self::data_ref`]
+    fn allows_data_ref(&self) -> bool;
+
+    /// Returns the underlying data range as a slice of bytes.
+    ///
+    /// # Panics
+    /// If the reader does not support returning the data range as a slice of bytes.
```

Comment:
this must reference the function above and the caller must ensure that this type supports data_ref before calling this